modules/renter/download.go

```diff
@@ -63,20 +62,22 @@ type (
 		atomicDataReceived uint64
 		downloadComplete   bool
 		downloadErr        error
-		finishedChunks     []bool
+		finishedChunks     map[int]bool
+		offset             uint64
+		length             uint64
 
 		// Timestamp information.
 		completeTime time.Time
 		startTime    time.Time
 
 		// Static information about the file - can be read without a lock.
 		chunkSize         uint64
-		destination       string
+		destination       modules.DownloadWriter
 		erasureCode       modules.ErasureCoder
 		fileSize          uint64
 		masterKey         crypto.TwofishKey
 		numChunks         uint64
-		pieceSet          []map[types.FileContractID]pieceData
+		pieceSet          map[int]map[types.FileContractID]pieceData
```

Comment:
The array type works as all chunks were downloaded (e.g. numbered 0 through N). Since `offset+length` downloads require a subset of chunks to be downloaded, a map allows tracking only the chunks that are actually required.